### ISSUE TITLE
[P2] Seed and record every stochastic source in the T5 training pipeline

### ISSUE THREAD

**Issue by ankaggarwal94** [ankaggarwal94/qanta-buzzer#40]
## Summary

The T5 supervised/PPO path exposes a split seed but does not seed the stochastic operations that determine model initialization, question order, rollout sampling, action sampling, or PPO mini-batches.

At current `main` (`5cafae3c4fc3dfd1525975aa34fda77975b7f4cb`):

- `configs/t5_policy.yaml` contains `data.seed`, which is used for dataset splitting.
- `SupervisedTrainer.train_epoch()` calls global `random.shuffle()`.
- `PPOTrainer.collect_rollouts()` calls global `random.choices()`.
- `PPOTrainer.update_policy()` calls `np.random.permutation()`.
- T5 model initialization and policy sampling use PyTorch RNGs.
- The entrypoint does not call `random.seed`, `numpy.random.seed`, `torch.manual_seed`, or CUDA seed controls before model construction/training.
- `config_used.json` and `split_manifest.json` do not record a complete training RNG/determinism contract.

Repeated runs with the same documented config and data split therefore do not identify the same experiment.

## Relevant code

- [`configs/t5_policy.yaml`](https://github.com/ankaggarwal94/qanta-buzzer/blob/5cafae3c4fc3dfd1525975aa34fda77975b7f4cb/configs/t5_policy.yaml)
- [`scripts/train_t5_policy.py`](https://github.com/ankaggarwal94/qanta-buzzer/blob/5cafae3c4fc3dfd1525975aa34fda77975b7f4cb/scripts/train_t5_policy.py)
- [`training/train_supervised_t5.py`](https://github.com/ankaggarwal94/qanta-buzzer/blob/5cafae3c4fc3dfd1525975aa34fda77975b7f4cb/training/train_supervised_t5.py)
- [`training/train_ppo_t5.py`](https://github.com/ankaggarwal94/qanta-buzzer/blob/5cafae3c4fc3dfd1525975aa34fda77975b7f4cb/training/train_ppo_t5.py)

## Minimal reproduction

Run the same small training configuration twice from clean checkpoint directories:

```bash
python scripts/train_t5_policy.py --config configs/t5_policy.yaml --smoke
mv checkpoints /tmp/checkpoints-run-1

python scripts/train_t5_policy.py --config configs/t5_policy.yaml --smoke
mv checkpoints /tmp/checkpoints-run-2
```

Even with identical split manifests, the shuffled batches, sampled rollout questions/actions, parameter initialization, and update order are not pinned. Checkpoint hashes and metrics can differ without any provenance field explaining the difference.

## Impact

- Seed-level variance cannot be distinguished from accidental non-reproducibility.
- A reported rerun with the same YAML is not a byte- or behavior-identical rerun.
- Comparing hazard warm-start, supervised-only, and PPO variants can confound treatment effects with uncontrolled RNG state.
- Debugging a regression from one checkpoint is needlessly difficult because the exact stochastic trajectory cannot be reconstructed.

This does not imply full GPU bitwise determinism is always available. It does require an explicit, honest contract about what is seeded and what remains nondeterministic.

## Expected behavior

Add a dedicated training seed contract, separate from the data-split seed, and initialize all relevant RNGs before model construction and any data sampling.

Recommended fields:

```yaml
training:
  seed: 13
  deterministic_algorithms: false
  cudnn_deterministic: false
  cudnn_benchmark: true
```

Use local RNG objects where practical rather than process-global state, and record effective device/library determinism information in run metadata.

## Acceptance criteria

- [ ] Add explicit model/training/rollout seeds; do not overload the split seed silently.
- [ ] Seed Python `random`, NumPy, PyTorch CPU, and all CUDA devices before model initialization.
- [ ] Use seeded/local RNGs for supervised shuffling, rollout question sampling, action sampling where supported, and mini-batch permutations.
- [ ] Persist seeds plus effective determinism flags, device, CUDA/cuDNN versions, and known nondeterministic exceptions in run metadata.
- [ ] A deterministic CPU smoke test produces identical sampled question sequences, batch orders, action draws, metric histories, and checkpoint tensors across two clean runs.
- [ ] A changed training seed changes at least one false-oracle trace.
- [ ] Multi-seed experiments namespace outputs by seed and never overwrite each other.
- [ ] Resume/checkpoint loading restores RNG state if exact continuation is claimed; otherwise label resumed trajectories as newly seeded.
- [ ] Document that GPU bitwise identity may not be guaranteed unless deterministic algorithms are enabled, while preserving statistical/run provenance.

## Validation plan

Instrument the trainers to emit the first N supervised QIDs, rollout QIDs, action draws, and mini-batch index permutations. Two CPU runs under the same seed must match exactly; a different seed must diverge. Then compare checkpoint state dicts for the deterministic smoke fixture.

**Comment by ankaggarwal94** [comment]
## Independent CoVe revalidation

**Verdict:** CONFIRMED at `main` `5cafae3c4fc3dfd1525975aa34fda77975b7f4cb`  
**Severity:** P2 retained  
**Confidence:** high

The T5 path has a data-split seed but no executed training-seed initialization before model construction. Independent source traces found all of the stochastic sources named in the issue:

- global `random.shuffle()` in supervised batching;
- global `random.choices()` in PPO rollout-question sampling;
- global `np.random.permutation()` in PPO mini-batch ordering;
- randomly initialized policy-head parameters and dropout under PyTorch RNG;
- `torch.distributions.Categorical` action sampling;
- no `random.seed`, `np.random.seed`, `torch.manual_seed`, `torch.cuda.manual_seed_all`, or equivalent `transformers.set_seed` call in the entrypoint/trainers.

`config_used.json` and `split_manifest.json` record the data/config snapshot, not Python/NumPy/PyTorch RNG state or a determinism contract.

**Disconfirmation/mitigation found:** the pretrained T5 encoder weights themselves are loaded from a fixed model identifier/checkpoint, but the policy heads, sample order, rollout choices/actions, dropout, and update order remain uncontrolled. Full GPU bitwise identity is not implied by ordinary seeding; the issue already distinguishes seeding/provenance from deterministic-algorithm guarantees.

One acceptance criterion should be read conditionally: byte-identical CPU checkpoint tensors require deterministic backend settings in addition to seed initialization. Trace identity and changed-seed divergence are the more portable false oracles.

No matching open or closed issue was found, and the current T5 tests contain no seed-replay regression.

Two full training runs were not executed in this connector-only audit; the absence of RNG initialization and presence of multiple stochastic calls directly establish non-reproducibility of the documented run contract.
